staging/src/k8s.io/apiserver/pkg/endpoints/handlers/delete_test.go

```diff
@@ -215,6 +219,205 @@ func TestDeleteCollection(t *testing.T) {
+			provider:                 true,
+			providerHasUnsafeDeleter: true,
+			authorizer:               &fakeAuthorizer{decision: authorizer.DecisionAllow, reason: "permitted"},
+			statusCode:               http.StatusOK,
+			unsafeDeleterInvoked:     1,
+		},
+	}
+
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			featuregatetesting.SetFeatureGateDuringTest(t, utilfeature.DefaultFeatureGate, features.AllowUnsafeMalformedObjectDeletion, test.featureEnabled)
+
+			info, ok := runtime.SerializerInfoForMediaType(codecs.SupportedMediaTypes(), runtime.ContentTypeJSON)
+			if !ok {
+				t.Fatalf("failed to setup serializer")
+			}
+			scope := &RequestScope{
+				Namer:      &mockNamer{},
+				Serializer: runtime.NewSimpleNegotiatedSerializer(info),
+				Authorizer: test.authorizer,
+			}
+
+			// keep track of what the normal and unsafe deleter observes
+			normalDeleterGot, unsafeDeleterGot := &holder{}, &holder{}
+
+			var deleter rest.GracefulDeleter = &fakeDeleter{holder: normalDeleterGot}
+			switch {
+			case test.provider && test.providerHasUnsafeDeleter:
+				deleter = &corruptObjDeleterProvider{
+					fakeDeleter:   &fakeDeleter{normalDeleterGot},
+					unsafeDeleter: &fakeDeleter{unsafeDeleterGot},
+				}
+			// provider, but it return nil
+			case test.provider:
+				deleter = &nilProvider{fakeDeleter: &fakeDeleter{normalDeleterGot}}
+			}
+
+			handler := DeleteResource(deleter, true, scope, &fakeAdmitter{})
+
+			req, err := http.NewRequest("GET", "/test", ioutil.NopCloser(strings.NewReader("")))
+			if err != nil {
+				t.Fatalf("unexpected error: %v", err)
+			}
+			reqInfo := &request.RequestInfo{IsResourceRequest: true}
+			req = req.WithContext(request.WithRequestInfo(req.Context(), reqInfo))
+			if test.ignore {
+				q := req.URL.Query()
+				q.Add("ignoreStoreReadErrorWithClusterBreakingPotential", "true")
+				req.URL.RawQuery = q.Encode()
+			}
+
+			recorder := httptest.NewRecorder()
+			handler.ServeHTTP(recorder, req)
+			resp := recorder.Result()
+
+			if want, got := test.statusCode, resp.StatusCode; want != got {
+				t.Errorf("expected status code: %d, but got: %d", want, got)
+			}
+			body, err := io.ReadAll(resp.Body)
+			if err != nil {
+				t.Fatalf("unexpected error reading response body: %v", err)
+			}
+			if want := test.errshouldContain; !strings.Contains(string(body), want) {
+				t.Errorf("expected response body to contain %q, but got: %s", test.errshouldContain, body)
+			}
+			if want, got := test.regularDeleterInvoked, normalDeleterGot.invoked; want != got {
+				t.Errorf("expected normal deleter to be nvoked %d times, but got: %d", want, got)
+			}
+			if want, got := test.unsafeDeleterInvoked, unsafeDeleterGot.invoked; want != got {
+				t.Errorf("expected unsafw deleter to be nvoked %d times, but got: %d", want, got)
+			}
+			// when invoked, the normal deleter should never see the ignore option enabled
+			if test.regularDeleterInvoked == 1 && normalDeleterGot.ignore != nil {
```

Comment:
fwiw, `normalDeleterGot.ignore != nil` could be false and we would still produce an error. It would be better to check for `!ptr.Deref[bool](unsafeDeleterGot.ignore, false)`

Reply:
I fixed it by initializing it to `true` and after the Delete is invoked we expect it to be reset to `nil`